zerver/tests/test_tusd.py

```diff
@@ -0,0 +1,291 @@
+        assert settings.MAX_FILE_UPLOAD_SIZE is not None
+        file_id = "e4b4acc5ddb8675d3af4e75a1a095bd7"
+        path_id = f"tusd/{file_id}"
+        body = {
+            "Upload": {
+                "ID": f"{file_id}+xyz",
+                "IsFinal": False,
+                "IsPartial": False,
+                "MetaData": {
+                    "filename": "brijsiyag.zip",
+                    "filetype": "application/zip",
+                    "name": "brijsiyag.zip",
+                    "type": "application/zip",
+                },
+                "Offset": 0,
+                "PartialUploads": None,
+                "Size": settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024 - 100,
+                "SizeIsDeferred": False,
+                "Storage": {"Type": "s3store", "Key": path_id},
+            },
+        }
+
+        upload_content_to_s3(bucket, f"tusd/{file_id}", None, user, b"zulip!")
+
+        result = self.client_post(
+            "/api/internal/tusd",
+            body,
+            content_type="application/json",
+            HTTP_HOOK_NAME="pre-finish",
+        )
+        self.assert_json_success(result)
+
+        result = self.client_get(f"/json/user_uploads/{file_id}")
+        result_data = self.assert_json_success(result)
+
+        upload_base = "/user_uploads/"
+        url = result_data["url"]
+        path_id = url[len(upload_base) :]
+
+        content = bucket.Object(path_id).get()["Body"].read()
+        self.assertEqual(b"zulip!", content)
+
+    def test_local_file_move_fails(self) -> None:
+        self.login("hamlet")
+        file_id = "e4b4acc5ddb8675d3af4e75a1a095bd7"
+        assert settings.LOCAL_UPLOADS_DIR is not None
+        file_path = os.path.join(settings.LOCAL_UPLOADS_DIR, "tusd", file_id)
+        body = {
+            "Upload": {
+                "ID": file_id,
+                "IsFinal": False,
+                "IsPartial": False,
+                "MetaData": {
+                    "filename": "zulip.txt",
+                    "filetype": "application/zip",
+                    "name": "zulip.txt",
+                    "type": "application/zip",
+                },
+                "Offset": 0,
+                "PartialUploads": None,
+                "Size": 100,
+                "SizeIsDeferred": False,
+                "Storage": {"Type": "filestore", "Path": file_path},
+            },
+        }
+        with self.assertLogs(level="ERROR") as error_log:
+            result = self.client_post(
+                "/api/internal/tusd",
+                body,
+                content_type="application/json",
+                HTTP_HOOK_NAME="pre-finish",
+            )
+        self.assertEqual(
+            error_log.output,
+            [
+                f"ERROR:root:error moving tusd file {file_path}: [Errno 2] No such file or directory: '{file_path}'"
+            ],
+        )
+        self.assert_json_error(result, msg="Upload failed.")
```

Comment:
I'm not sure I understand the test here -- is the point that we didn't actually upload the file and then tried to run the `pre-finish` hook?

Reply:
I've not fixed the tests in any way yet.